src/System.Linq.Expressions/src/System/Dynamic/Utils/DelegateHelpers.cs

```diff
@@ -53,101 +106,173 @@ private static Delegate CreateObjectArrayDelegateRefEmit(Type delegateType, Func
                 bool hasReturnValue = returnType != typeof(void);
 
                 ParameterInfo[] parameters = delegateInvokeMethod.GetParametersCached();
-                Type[] paramTypes = new Type[parameters.Length + 1];
-                paramTypes[0] = typeof(Func<object[], object>);
-                for (int i = 0; i < parameters.Length; i++)
-                {
-                    paramTypes[i + 1] = parameters[i].ParameterType;
-                }
 
-                thunkMethod = new DynamicMethod("Thunk", returnType, paramTypes);
-                ILGenerator ilgen = thunkMethod.GetILGenerator();
+                do
+                {
+                    if (parameters.Length > 2)
+                    {
+                        break; // Don't use C# thunks for more than 2 parameters
+                    }
 
-                LocalBuilder argArray = ilgen.DeclareLocal(typeof(object[]));
-                LocalBuilder retValue = ilgen.DeclareLocal(typeof(object));
+                    if (returnType.IsByRefLike || returnType.IsByRef || returnType.IsPointer)
+                    {
+                        break; // Don't use C# thunks for types that cannot be generic arguments
```

Comment:
What about other types like `ArgIterator`? Those are illegal as C# type arguments and IIRC they're also illegal at the runtime level. Do they need to be handled here?

Reply:
Sigh. I thought that (and TypedReference) were marked as ByRefLike types. It looks like they aren't though. Well, I think those cases are sufficiently rare that I can fix that by simply wrapping the instantiation of the generic method within a try/catch, and falling back to the dynamic method case. The performance of that will be terrible, but its such a rare scenario, that I believe the exception path is reasonable.

Reply:
Well, that, and TypedReference are handled by the IsByRefLike flag. However, in general, I can't easily write checks for function pointers or other unknown types with special semantics, so I'm wrapping this whole thing in an unconditional try/catch which will fall back to using the normal dynamic method path.

Reply:
Sounds like a plan.

---
In reply to: [340323867](https://github.com/dotnet/corefx/pull/42187#discussion_r340323867) [](ancestors = 340323867)